### ISSUE TITLE
Wapcaplet's Widget Wishlist Woadmap

### ISSUE THREAD

**Issue by wapcaplet** [CleverRaven/Cataclysm-DDA#53957]
### Is your feature request related to a problem? Please describe.

Back when I started working on sidebar widgets in #44683, I hoped to eventually make the entire HP/stats/log sidebar "moddable", that is, defined and laid out entirely with JSON declarations, rather than hard-coded, fixed-width display functions.

After [many other PRs](https://github.com/CleverRaven/Cataclysm-DDA/pulls?q=is%3Apr+author%3Awapcaplet+is%3Amerged+widget+), my dreams are finally coming to fruition in #51438.

But the journey isn't over, and there are still things to do before this rather game-changing feature can be considered complete.

I am opening this issue to share my planned future direction for widgets, and invite others to collaborate on testing and improving the system before 0.G.

### Solution you would like.

Immediate:
- Error handling - with wrong or unexpected fields in widget JSONs, what happens? I mostly tested only with valid JSON definitions, but this is expected to be strongly user-facing, so it's essential to have helpful messages when you forget the "var" field or try to use "number" style with a text variable, etc.
- Moddability - is it possible for mods to add sidebars? why would they want to?
- Playability - does a pure widget sidebar provide enough info to play the game?

Near-term (before 0.G):
- Widgets for status effects (only found in @ now)
- Widgets for speed and penalties (only found in @ now)
- Widgets for armor coverage, layers, total protection
- Multi-bodypart widgets in a table, to help condense per-bodypart attributes (HP, encumbrance, warmth, wetness)

Long-term (after 0.G):
- Widgets for skill levels, bionics, proficiencies, traits
- Widgets for character height, weight, age, hair color
- Widgets for friendly NPC stuff (name, location, wielded weapon, activity/assignment)
- Dynamic layouts or sidebars that can change in response to in-game effects, mutations, etc.
-- New bodypart widget when growing a tail
-- New faction camp ally widget when befriending an NPC

### Additional context

References:
- [doc/SIDEBAR_JSON.md](doc/SIDEBAR_JSON.md)
- [My merged widget-related PRs](https://github.com/CleverRaven/Cataclysm-DDA/pulls?q=is%3Apr+author%3Awapcaplet+is%3Amerged+widget+)

**Comment by PatrikLundell** [comment]
I can imagine that a mod might want to display remaining duration for a number of things where the player is expected to know the duration, such as e.g. spells. For things where the player is expected to know the approximate duration it might also be useful to have widgets with suitable vague terms to indicate the duration (it's probably a bad idea to have a "flight" spell expire while high up in the air, for instance).
Another use might be the display of remaining resources, such as bionic power, mana, etc.
I don't think widgets for static info such as height and hair color (yes, it can be changed, but it's generally done infrequently) is useful. If someone has a triple monitor setup they might want to allocate unused space for "extra" info, especially if widgets can be placed in spaces that would otherwise remain unused.
Weight is slightly more useful (but you'll already know you're too heavy to ride a pony when geared up), but it's available only when you have a scale...

- I can see proficiencies and skills being of interest while practicing them, to see the progress.
- Armor coverage, etc. might be used temporarily while trying out various gear combinations.
- An indicator widget that tells you your armor has been damaged could be useful to remind you that you need to repair your gear, but that would either require some memory, or indicate damage whenever something is damaged, even if it was damaged before you put it on. However, even an indicator that just shows whether anything is damaged would be useful.

Some of these indicated widget thingies might be better done elsewhere, rather than squeezed into the widget framework. Just because it can be widgetisized doesn't mean it's a good idea to do so, or that it's the best way to provide that info...

**Comment by wapcaplet** [comment]
Remaining duration for spells is a great idea - this is a case where a more dynamic section will be important, so it can adapt to showing whichever spell(s) you have recently cast.

Good point about the relatively static character attributes; I thought of those because they'd be pretty easy to do, but you're right that they probably aren't very useful. For weight, there is already the `weight_desc` one for "Normal", "Overweight", etc. which is probably good enough.

Thanks for the feedback, I will keep your suggestions in mind.

**Comment by wapcaplet** [comment]
The original intention was for widgets to make the sidebar moddable, but I never tried it with a true mod until now. It was pretty easy.

Commit adding the mod: https://github.com/wapcaplet/Cataclysm-DDA/commit/e0d4ab29b4c5a85ba13163a5d921f834b070ebc6

As a simple demo, this "SelfAwareSidebar" mod adds two layouts showing numbers for HP, stats, health, and pain, as an homage to our dearly departed Self-Aware trait (#53802). A slim 16-character-wide version:

![image](https://user-images.githubusercontent.com/238395/148630215-5475292c-bb1f-432d-b3b3-b59aa63d903c.png)

And an extra-wide 60-character version:

![image](https://user-images.githubusercontent.com/238395/148630221-9d84f060-dace-4e7d-82e6-7f6d2d71009c.png)

This is just a proof of concept - showing that mods can easily define their own widgets, layouts, and sidebars.

**Comment by Inglonias** [comment]
I did a little bit of messing around with a custom sidebar mod this morning, and while I didn't get too far, one thing I wish I had was a list of all of the already existing stock widgets somewhere. The last thing I want to do is spend time designing and debugging my own widgets when the game already comes with a perfectly good implementation of what I want. I guess the thing I want here is a piece of documentation that has a list of all stock variable widgets, layout widgets, and "screenshots" of said layout widgets, even if those screenshots are just pre-formatted text blocks representing what they'd look like.

**Comment by wapcaplet** [comment]
> I guess the thing I want here is a piece of documentation that has a list of all stock variable widgets, layout widgets, and "screenshots" of said layout widgets, even if those screenshots are just pre-formatted text blocks representing what they'd look like.

Good idea, I opened #54186 to work on adding some of these to the docs.

**Comment by Inglonias** [comment]
Wanted to update this issue - #58113 supersedes #54191.
